TSL: Add CDLNode

donmccurdy opened the pull request:
Implements ASC Color Decision List (CDL) v1.2.

I'm not ready to do an integration with the TSL tone mapping nodes quite yet, but here's an early comparison, configuring the CDL to emulate the AgX “Punchy” look, with higher saturation and contrast:

| ACES Filmic | Reinhard | AgX Base | AgX Punchy |
|-------------|----------|----------|------------|
|  ![coffeemat_aces](https://github.com/user-attachments/assets/876b60b4-8fc3-4c7c-baa4-a99b0b668371) | ![coffeemat_reinhard](https://github.com/user-attachments/assets/231f6a7e-9180-4fef-98f5-39770c6e433f) | ![coffeemat_base](https://github.com/user-attachments/assets/66538624-c6c5-4db5-99cf-af040e64f3b5) | ![coffeemat_punchy](https://github.com/user-attachments/assets/0e167b76-4c85-4711-93a9-e757a48764b3) |

The goal is to provide configurability for our current hard-coded tone mappers.

Related:

- #28544

/cc @WestLangley @sunag

donmccurdy committed:
TSL: Add CDLNode

WestLangley commented on src/nodes/display/CDLNode.js:
```diff
@@ -0,0 +1,83 @@
+
+	}
+
+	constructor( inputNode, slopeNode, offsetNode, powerNode, saturationNode ) {
```
Is it overkill for the parameters to be Nodes? Can't they just be float (or ultimately vec3)

donmccurdy replied:
Probably right – I'll make the change to just constant `float | vec3` inputs unless @sunag has other preferences for node best practices? 

We should give each a default value, as well.

WestLangley replied:
I'd really like to see them as vec3 so we can experiment.

sunag replied:
I don't think it would be a problem, as you are using `nodeObject()` in the `cdl()` function parameters, it can support any primitive value like a JS number, constant, uniform or procedural, float or vec3. We can let the user choose what is best for him.

WestLangley replied:
@donmccurdy Maybe you can use node args and write the code it works with either float or vec3.

donmccurdy replied:
@WestLangley it does work with either float or vec3. :) 

With vec3 input we can, for example, do the "AgX Golden" look:

![chair_agx_golden](https://github.com/user-attachments/assets/ff3f8a25-1d44-41c6-b142-1bb3eea7d0c3)

WestLangley replied:
Great! That is exactly why I was pushing for vec3 support!

WestLangley commented on src/nodes/display/CDLNode.js:
```diff
@@ -0,0 +1,83 @@
+			const luma = inputNode.rgb.dot( vec3( luminanceCoefficients ) );
+
+			// clamp( ( in * slope ) + offset ) ^ power
+			const output = max( inputNode.rgb.mul( slopeNode ).add( offsetNode ), 0.0 ).pow( powerNode ).toVar();
```
As we discussed offline, this should be implemented in log space, and then converted back to linear.

See https://www.mdpi.com/2313-433X/3/4/40

This can be corrected later if you want.

donmccurdy replied:
I have some doubts about this rule (see introduction to https://blender.stackexchange.com/a/55239/43930), but, all usage I have in mind for future PRs would indeed be within a log space. That said — I don't think we can put the conversion inside of CDLNode, as the particular choice of log space may vary. If we're implementing looks for AgX we'd likely use AgX Log (Kraken), and for ACES Filmic we'd perhaps use ACEScc. #29450 may eventually need to be extended to handle the common log spaces.

WestLangley replied:
I was basing my comment on Filament's [implementation](https://github.com/google/filament/blob/main/filament/src/details/ColorGrading.cpp) and the assumption that log space was the correct space. I am not sure if the particular choice of log space is important for our purposes.

Since we have a node-based system -- and not a fixed pipeline -- I would convert to log and back inside the CDLNode so the calculation is correct.

donmccurdy replied:
The CDL spec and the ACES document above disagree about this; the former is very clear that application in linear spaces is allowed, so long as you're consistent in communicating the color space expected for any particular set of CDL coefficients.

The ACES pipeline can of course be more opinionated about what to allow, as can we... But I don't think we can embed a particular color space transform into the operation and still call it a CDL, it will be impossible to compare results with other software, and that's the real point of using a CDL. 

If we are offering an artist-friendly color grading API like Filament’s ColorGrading.cpp (perhaps we should!), then embedding the log conversion would be entirely appropriate. The CDL would be a single step, in log space, within that process. But the CDL is a mathematical operation, low-level and not particularly artist-friendly, and it must be possible to compose different pipelines around it... which I feel that an internal log conversion will block.

donmccurdy replied:
Sorry, I'm afraid don't see how this can work. Other common color operators (like contrast) may receive comparable benefit from using a log space, and I think it will be impractical (for performance, numerical precision, and creative intent) to enforce a color space conversion within low-level operators.

Note that the CDL is not itself a "Look" (or LMT as in the ACES literature), a CDL might be just a component of a look. One CDL might be applied before tone mapping, and another after. For that reason, we also cannot require that the input to a CDL must be in the global working color space.

To your points — if we wanted to say that a _Look_ or _LMT_ should always be done in a log space, I'm aware of no issues with that statement, and I have no plans to do otherwise!

WestLangley replied:
>The CDL spec ...  is very clear that application in linear spaces is allowed

You are correct. Both linear and log space are allowed.

Just so I understand your point, are you planning to add a log-conversion Node of some sort, so users who prefer to apply the CDL in log space can do so?

donmccurdy replied:
> are you planning to add a log-conversion Node of some sort ...

Yes, certainly! Or maybe log conversion could be supported by THREE.ColorManagement and the existing ColorSpaceNode directly... but I'm nervous about getting that right, and inclined to start with 1-2 dedicated log space nodes instead.

I am also hoping to find a simpler higher-level abstraction so that most users do not need to interact with the CDL node directly, but I am not sure, yet, what that will mean.

WestLangley replied:
OK. I see we are not quite sure where this is headed... So I think this is OK, and we'll see what happens!  :-)

An example would be awesome. 🙏

sunag commented on src/nodes/display/CDLNode.js:
```diff
@@ -0,0 +1,83 @@
+
+export default CDLNode;
+
+export const cdl = ( node, slope, offset, power, saturation ) => nodeObject(
```
I think can add the default values ​​here, this helps with `typing` and since the process is before of `nodeObject()` we can use JS numbers, nodes or native three.js `Vector*` classes.

donmccurdy committed:
CDLNode: Add default vec3 inputs

donmccurdy committed:
CDLNode: Add default vec3 inputs to cdl()

donmccurdy commented:
@sunag I notice the nodes defined in https://github.com/mrdoob/three.js/blob/dev/src/nodes/display/ColorAdjustment.js are simpler, just a `Fn` declaration and not a class. Are there particular pros/cons to defining a node one way or another? I'm tempted to switch this over to a `Fn` declaration and moving it there.

sunag commented:
Sounds like a good idea, if you use `Fn` this will simplify it, it will automatically apply `nodeObject()` to the parameters and reduce the initial structure.

sunag commented:
I think we can move this node for addons too? https://github.com/mrdoob/three.js/pull/29511

donmccurdy commented:
I think my preference would be to move it into `ColorAdjustment.js` — especially once it's simplified, it's not so different from the other functions there. But moving the color adjustment functions all into addons could make sense, perhaps excluding luminance, which I think we depend on elsewhere.

donmccurdy committed:
move cdl() function to ColorAdjustments.js

donmccurdy commented:
Reimplemented as a simpler function and moved to `ColorAdjustment.js`. I've also changed the clamp closer to the approach taken by Filament, though I'm unsure whether we need the final `max(output, 0.0)` step.

sunag commented:
> I'm unsure whether we need the final max(output, 0.0) step.

I think it would be better to use `max` externally if necessary since the limits can be applied easily.

It would be better remove the names from `.toVar()`, this may conflict with other variables depending on the flow there may be other names that are the same, it is certainly something we will improve soon.

donmccurdy committed:
clean up cdl(), add more JSDoc

donmccurdy commented:
@sunag done and done! I did find `.toVar( 'name' )` very useful while debugging an error in my shaders, so if that's something we can make easier in the future that would be awesome. :)

donmccurdy committed:
clean up

donmccurdy committed:
clean up
